src/misc/misc.3
.\"(c) Copyright 1992 by Panagiotis Tsirigotis
.\"All rights reserved.  The file named COPYRIGHT specifies the terms 
.\"and conditions for redistribution.
.\"
.\" $Id: misc.3,v 1.1 2001/05/26 22:04:49 seth Exp $
.TH MISC 3X "25 January 1992"
.SH NAME
make_string, make_pathname, argv_copy_and_clear, basename, dirname - miscellaneous functions
.SH SYNOPSIS
.LP
.nf
.ft B
#include "misc.h"
.LP
.ft B
char *make_string( count, ... )
unsigned count ;
.LP
.ft B
char *make_pathname( count, ... )
unsigned count ;
.LP
.ft B
char **argv_copy_and_clear( argv, start, count )
char **argv ;
int start, count ;
.LP
.ft B
char *dirname( path )
char *path ;
.LP
.ft B
char *basename( path )
char *path ;
.SH DESCRIPTION
.LP
This library contain miscellaneous functions, hence the name.
.LP
.B make_string()
creates a string by catenating the list of strings passed as arguments.
\fIcount\fR indicates how many strings there are.
Strings that are
.SM NULL
pointers are ignored.
\fBmake_string()\fR returns malloc'ed memory.
.LP
.B make_pathname()
creates a pathname by catenating the list of pathname components passed
as arguments and inserting slashes between them.
\fIcount\fR indicates how many components there are.
\fBmake_pathname()\fR returns malloc'ed memory.
.LP
.B argv_copy_and_clear()
creates a new argv array of size \fIcount\fR, and fills it with the
contents of \fIargv\fR from \fIstart\fR up to \fIstart+count-1\fR.
The original \fIargv\fR entries in that range are cleared by filling
them with
.SM SPACEs.
.LP
.B dirname()
returns in a malloced string containing all but the last of 
component of \fIpath\fR. There are 2 special cases:
first when the
\fIpath\fR is "/", \fBdirname()\fR will return "/", 
and second, when the \fIpath\fR does not contain any '/',
\fBdirname()\fR will return ".".
.LP
.B basename()
returns a pointer to the last component of \fIpath\fR.
.SH "RETURN VALUES"
.LP
\fBmake_string()\fR returns a pointer to the new string.
It returns
.SM NULL
if \fIcount\fR is 0.
.LP
\fBmake_pathname()\fR returns a pointer to the new pathname.
It returns
.SM NULL
if \fIcount\fR is 0.
.LP
\fBargv_copy_and_clear()\fR returns the new argv array. It returns
.SM NULL
if \fImalloc\fR fails to allocate more memory.
.LP
\fBdirname()\fR returns a new string or
.SM NULL
if \fImalloc\fR fails to allocate more memory.
.LP
\fBbasename()\fR returns a pointer to the last component of \fIpath\fR.
.SH BUGS
.LP
The behavior of \fBdirname()\fR and \fBbasename()\fR is undefined if
they are given a zero-length string.


ChangeLog
2003-07-16  Seth Robertson  <seth@startide.sysdetect.com>

	* src/fsma/fsma.c (_fsm_free): Prevent infinite loops on fsm_free
	of corrupted memory

2003-07-09  James Tanis  <jtt@hawking.sysdetect.com>

	* src/dict/dll.c (dll_error_reason): Export pulibcally

2003-06-13  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h, include/ht.h, include/pq.h: avoid stupid void * #1311

2003-06-07  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_create): argsp may change size between
	external and internal users due to HASH_STATS define.

2003-05-16  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_destroy): Allow clean up of all memory when
	insure is in use and fsma is replaced by malloc

2003-05-10  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h (DICT_STATS): keep performance stats for hash tables

	* src/dict/htimpl.h (struct ht_stats): hash table performance stats
	(ENTRY_IS_FULL): kludge to detect bucket collisions

	* src/dict/ht.c: instrumented to keep performance stats for hash tables
	(print_hashstats): print performance stats

2003-04-28  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/dict/ht.c (ht_delete): microscopic performance tweak
	[Seth got the big win already]

2003-04-16  James Tanis  <jtt@saint.sysdetect.com>

	* src/fsma/fsma.c (fsm_threaded_makeready): Only allow users to
	actually alter the value of thread_overried if we are compiled
	with threads.

2003-04-08  Seth Robertson  <seth@startide.sysdetect.com>

	* src/fsma/fsma.c (fsm_threaded_makeready): Provide interface for
	people to specify this will be a threaded program and to enforce
	thread safe memory protection.

2003-04-03  Seth Robertson  <seth@startide.sysdetect.com>

	* src/pq/hpq.c: Make thread-safe, through PQ_THREADED_SAFE

2003-03-31  Seth Robertson  <seth@winwood.sysdetect.com>

	* src/dict/*.c src/fsma/*.c: Make thread-safe, through
	DICT_THREADED_SAFE and DICT_THREADED_MEMORY

2002-12-10  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/pq/Makefile: move -lpq to -lpqueue so as not to conflict
	with postgres

2002-11-11  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/fsma/fsma.c, include/dict.h: opx_rollup_3 merge

2002-11-09  James Tanis  <jtt@stevedore.sysdetect.com>

	* include/dict.h, src/fsma/fsma.c: trheads merge

2002-11-06  Seth Robertson  <seth@stevedore.sysdetect.com>

	* src/fsma/fsma.c: Compile problems when using dmalloc

2002-11-05  Seth Robertson  <seth@stevedore.sysdetect.com>

	* include/dict.h: Define symbol for NOCOALESCE which makes it
	clear the only reason we are using it is because we are too lazy
	to make CLC thread-safe

2002-09-09  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/dict/ht.c: insight nits

2002-09-05  Seth Robertson  <seth@stevedore.sysdetect.com>

	* src/dict/dll.c, ChangeLog: Do not create dlls w/o UNORDERED and
	w/o OOCMP function--coredump otherwise sometimes

2002-09-05  Seth Robertson  <seth@winwood.sysdetect.com>

	* src/dict/dll.c (dll_create): Default order is ordered.

	* src/dict/dll.c (dll_create): Use bitwise and for bits, instead
	of addition.  Prevented an argument sanity check from occurring.

2002-08-29  Seth Robertson  <seth@startide.sysdetect.com>

	* include/fsma.h (struct __fsma_header): AHHHHH-HHHHAAAAAAAA!
	The reference counter was previously a short.  Worker, for reasons
	too complicated to get into, can have more than 2^16th lists.
	If we happen (not unlikely) to destroy down to 2^16 (e.g. zero)
	fsma thinks all allocators are free and promptly proceeds to return
	the memory to the system.  Memory was free'd but still in use.
	BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA.

	* src/fsma/fsma.c (fsm_destroy): Perform assertion to verify we
	actually found the fsma allocator in the coalesce list.  Don't
	bother to realloc the coalesce list size down--it won't do no good
	nohow.

2002-08-27  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_minimum): Serious and quite possibly
	bug-injecting misuse of a hint (cur_min).  The insert and delete
	routines were setting this to a bucket chain entry, not a table
	entry bucket (except when the insert routine had to perform a
	bucket chain expansion--when cur_min might be set to some other
	random value).  This usually was OK since bucket chain numbers are
	very low (0-3 range), but if a valid entry was in the table at
	slot zero or one, then ht_minimum might never see that item.

2002-08-14  James Tanis  <jtt@saint.sysdetect.com>

	* src/dict/ht.c (ht_minimum): Use cached minimum index as search start.
	(ht_do_insert): Update minimum index if less the current minimum.
	(ht_delete): Search for new minimum index if the index of the
	deleted node equals the current minimum.
	In addition add and obj_cnt so we don't try to perform expensive
	operations on empty hashes.

2002-05-13  James Tanis  <jtt@saint.sysdetect.com>

	* src/fsma/Makefile (BK_LARGE_LIB): Shared libfsma no longer
	breaks with libtool (on both linux and bsd).

2002-05-10  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/fsma/Makefile (BK_LARGE_LIB): back out shared libfsma;
	this breaks libtool in a strange way that requires investigation

2002-05-10  James Tanis  <jtt@winwood.sysdetect.com>

	* src/fsma/fsma.c: Formatting and simplify coalesce code.

	* src/fsma/Makefile (BK_LARGE_LIB): Make libfsma shared.

	* src/dict/bst.c (bst_redelete): Improve efficiency by avoiding
	unnecessary checks and recusive calls.

2002-02-22  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h (DICT_NUKE): delete contents and dict itself
	(_DICT_NUKE_WORK_LOOP): common code for DICT_NUKE, DICT_NUKE_CONTENTS

	* src/dict/dictimpl.h (HANDLE_ERROR): Eliminate unused id argument
	[use BK_FUNCNAME  here instead if you need it for debugging]

2002-01-31  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/dict/Makefile (BK_LARGE_LIB): use ${LIBEXT} for libtool build

2002-01-09  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/fsma/fsma.c: INSIGHT->INSURE


src/misc/misc.c
/*
 * (c) Copyright 1992 by Panagiotis Tsirigotis
 * All rights reserved.  The file named COPYRIGHT specifies the terms 
 * and conditions for redistribution.
 */

static const char RCSid[] = "$Id: misc.c,v 1.2 2002/07/18 22:52:48 dupuy Exp $";
static const char misc_version[] = VERSION;

#include <stdarg.h>
#include <stdlib.h>

/*
 * MISCELLANEOUS FUNCTIONS
 */

#include "misc.h"
#include "string.h"
#include "clchack.h"

#ifndef NULL
#define NULL		0
#endif



/*
 * Create a new argv array,
 * copy the original to the new one,
 * and clear the old one
 */
char **argv_copy_and_clear(char **org_argv, int start, int count)
	                 							/* original argv */
	           									/* from where to start copy/clear */
	           									/* how many entries to copy/clear */
{
	char **new_argv ;
	char *p ;
	int i ;
	int j ;
#ifndef __STDC__
  	char *malloc() ;
#endif

	new_argv = (char **) malloc( count * sizeof( char * ) ) ;
	if ( new_argv == NULL )
		return( NULL ) ;

	for ( i = 0 ; i < count ; i++ )
	{
		new_argv[ i ] = make_string( 1, org_argv[ start+i ] ) ;
		if ( new_argv[ i ] == NULL )
		{
			for ( j = i-1 ; j >= 0 ; j-- )
				free( new_argv[ j ] ) ;
			free( (char *) new_argv ) ;
			return( NULL ) ;
		}
		for ( p = org_argv[ start+i ] ; *p ; p++ )
			*p = ' ' ;
	}
	return( new_argv ) ;
}

#ifndef __linux__
/*
 * We always return a pointer in pathname
 */
char *basename(char *pathname)
{
	char *s = strrchr( pathname, '/' ) ;

	if ( s == NULL )
		return( pathname ) ;
	else
		return( s+1 ) ;
}
#endif /* !__linux__ */


/*
 * We always return a malloced string
 *
 * There are 2 special cases:
 *
 *		1) pathname == "/"
 *				In this case we return "/"
 *		2) pathname does not contain a '/'
 *				In this case we return "."
 */
char *dirname(char *pathname)
{
	int len ;
	char *s = strrchr( pathname, '/' ) ;
	char *p ;
#ifndef __STDC__
  	char *malloc() ;
#endif

	if ( s == NULL )
		return( make_string( 1, "." ) ) ;
	else
	{
		len = s - pathname ;
		if ( len == 0 )
			return( make_string( 1, "/" ) ) ;
	}

	p = (char *) malloc( len+1 ) ;
	if ( p == NULL )
		return( NULL ) ;
	else
	{
		strncpy( p, pathname, len )[ len ] = '\0' ;
		return( p ) ;
	}
}


char *make_string( unsigned count, ...)
{
	va_list ap ;
	register unsigned i ;
	register unsigned len = 0 ;
	register char *s, *p ;
	char *new_string ;

	if ( count == 0 )
		return( NULL ) ;

	va_start( ap, count ) ;
	for ( i = 0 ; i < count ; i++ )
	{
		s = va_arg( ap, char * ) ;
		if ( s == NULL )
			continue ;
		len += strlen( s ) ;
	}
	va_end( ap ) ;

	new_string = (char *) malloc( len + 1 ) ;
	if ( new_string == NULL )
		return( NULL ) ;

	p = new_string ;
	va_start( ap, count ) ;
	for ( i = 0 ; i < count ; i++ )
	{
		s = va_arg( ap, char * ) ;
		if ( s == NULL )
			continue ;
		while (( *p++ = *s++ )) ;
		p-- ;
	}
	va_end( ap ) ;
	return( new_string ) ;
}


char *make_pathname( unsigned count, ... )
{
	va_list ap ;
	register unsigned i ;
	register unsigned len = 0 ;
	register char *s, *p ;
	char *pathname ;

	if ( count == 0 )
		return( NULL ) ;

	va_start( ap, count ) ;
	for ( i = 0 ; i < count ; i++ )
	{
		s = va_arg( ap, char * ) ;
		len += strlen( s ) ;
	}
	va_end( ap ) ;

	pathname = (char *) malloc( len + count ) ;
	if ( pathname == NULL )
		return( NULL ) ;

	p = pathname ;
	va_start( ap , count) ;
	for ( i = 0 ; i < count ; i++ )
	{
		s = va_arg( ap, char * ) ;
		while (( *p++ = *s++ )) ;
		*(p-1) = '/' ;			/* change '\0' to '/' */
	}
	*(p-1) = '\0' ;
	return( pathname ) ;
}

